src/librustc_mir/interpret/intrinsics.rs

```diff
@@ -78,6 +78,25 @@ impl<'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>> InterpretCx<'a, 'mir, 'tcx, M>
                 let id_val = Scalar::from_uint(type_id, dest.layout.size);
                 self.write_scalar(id_val, dest)?;
             }
+
+            "type_name" => {
+                let const_val = type_name(self.tcx.tcx, substs.type_at(0)).val;
+                if let ConstValue::Slice {
+                    data: slice_val,
+                    start: _,
+                    end: _,
+                } = const_val
+                {
+                    // unsafe for from_utf8_unchecked()
+                    unsafe {
```

Comment:
please get rid of `str_to_immediate`, it is only used in miri's implementation of `type_name` and thus not needed.

Instead, you can write

```rust
let alloc_id = self.tcx.alloc_map.lock().create_memory_alloc(data);
let name_val = Immediate::new_slice(Scalar::Ptr(ptr.into()), data.len(), self);
```

The code as currently written would be invalid when encountering a `ConstValue::Slice` where `start != 0` or `end - start != data.len()`. The way `type_name` is written, that doesn't happen, so... idk, add some comments explaining why that's ok?

Reply:
Maybe it would also simplify things if you pulled out https://github.com/rust-lang/rust/blob/0f0ad45098e7203451da99f7df4bfdbe22c9edf9/src/librustc_mir/interpret/intrinsics/type_name.rs#L216-L219 into a separate function (with `pub(super)` visibility) and called that here? That way we'd not have the `ConstValue::Slice` wrapper and could simplify the code shown above to

```rust
let alloc_id = self.tcx.alloc_map.lock().create_memory_alloc(data);
let name_val = Immediate::new_slice(Scalar::Ptr(ptr.into()), data.len(), self);
```

Reply:
> 😆 the reddit post you linked to shows a typecast example

Yeah. As I think I stated in there though, I don't think those are particularly realistic use cases for what someone who is using compile-time intrinsics at all is likely to want to do with this kind of thing. That kind of thing had never even crossed my mind as a concern until it was brought up by a few people in that thread.

With regards to the changes, I'll work on them now (I didn't notice you had replied until right after I sent off the minor commit I just made, haha.)

Thanks for the review!